Implement the convenience methods count and collect in DataSet

mxm opened the pull request:
These methods provide convenience for the API user to get intermediate results into the program.

StephanEwen commented on flink-core/src/main/java/org/apache/flink/api/common/accumulators/ListAccumulator.java:
```diff
@@ -0,0 +1,97 @@
+	public void write(DataOutputView out) throws IOException {
+		ObjectOutputStream outStream = new ObjectOutputStream(new DataOutputViewStream(out));
+		outStream.writeObject(typeSerializer);
+		outStream.flush();
```
Let us change this to `close()`

StephanEwen commented on flink-java/src/main/java/org/apache/flink/api/java/DataSet.java:
```diff
@@ -302,7 +308,48 @@ public ExecutionEnvironment getExecutionEnvironment() {
+	 *
+	 * @see org.apache.flink.api.java.Utils.CountHelper
+	 */
+	public Long count() throws Exception {
```
I would make this a primitive `long`

StephanEwen commented on flink-java/src/main/java/org/apache/flink/api/java/Utils.java:
```diff
@@ -32,4 +38,44 @@ public static String getCallLocationName(int depth) {
+
+		@Override
+		public void flatMap(T value, Collector<Long> out) throws Exception {
+			getRuntimeContext().getLongCounter(id).add(1L);
```
I think it is more efficient to have a `long`inside this function that is incremented for each record, and then aggregate this `long` to the accumulator in the `close()` method.

StephanEwen commented on flink-runtime/src/test/java/org/apache/flink/runtime/AbstractIDTest.java:
```diff
@@ -23,8 +23,8 @@
 import org.junit.Test;
-
 import org.apache.flink.runtime.testutils.CommonTestUtils;
+import org.apache.flink.util.AbstractID;
```
Let us move this test to `flink-core`, with `AbstractID` being now part of the core.

mxm replied:
As far as I see, `AbstractIDTest` is already part of `flink-core`.

mxm replied:
Nevermind, I forgot to push the last changes I made..

StephanEwen commented:
With the scheduler and intermediate data set enhancements coming up for 0.9 soon, this is now quite feasible to use. I suggest to merge it once the inline comments are addressed.

mxm commented:
I fixed the proposed changes and rebased to the current master. The changes are only reflected in the Java API and need to be added to the Scala API as well.

mxm commented:
I've implemented count and collect in the Scala API. There is still a problem with the `ListAccumulator` for non-primitive Objects (e.g. not Integer or Long) probably due to Object reuse.

mxm commented:
Looks like this is now ready to merge. 

@zentol I understand your concern. However, I think that it is much easier to execute in this way. Most of the times, the user probably wants just one accumulator result and not multiple. This is supposed to be a convenience function.

StephanEwen commented:
@zentol  You are right, for the time being, that this results in parts in repeated execution. While not totally unavoidable in all cases, the code going in soon about caching intermediate results will help there big time.

mxm commented:
I squashed the commits and rebased to the current master. Any objections against merging this?

rmetzger commented on flink-core/pom.xml:
```diff
@@ -58,6 +58,12 @@ under the License.
 			<version>0.5.1</version>
 		</dependency>
 
+		<dependency>
```
Why are you moving netty to `flink-core` ?

mxm replied:
Simply because AbstractID depends on `io.netty.buffer.ByteBuf`

rmetzger replied:
Allright, that makes sense

uce replied:
Actually, only a single convenience method in `AbstractID` depends on this. I would propose to refactor that method into one of the existing utility classes in `flink-runtime` and keep the Netty dependency in `flink-runtime`.

StephanEwen commented:
Let us not move netty to flink-core. Let us rather pull that method out of the AbstractID, or, not use AbstractID at in your code, but the Java UUID.

mxm commented:
I moved the `writeInto(ByteBuf buf)` method to the subclasses of `AbstractID` used by `NettyMessage`. This is the only context where the method is being used and the subclasses already have another `fromByteBuf(ByteBuf buf)` method.

mxm commented:
If there are no objections, I will merge this once Travis passes.
